neps/nep-0488.md

```diff
@@ -0,0 +1,1537 @@
+
+The draft implementation to nearcore you can find by this link[^54]. This implementation is based on *blst* library[^30]. This library one of the fastest[^45] and the most audited[^55].
+
+## Security Implications
+
+The implementation security relies on the security of the chosen library, which supports operations with BLS curves.
+
+In this NEP, we do not require a constant execution time for all operations. This is not a problem if you use host functions to verify the BLS signature. These host functions should not be used if you need a constant-time algorithm.
+
+The BLS12-381 has more security bits than the already existing pairing-friendly curve BN254, as a result, the security of the projects which need the pairing-friendly curve will improve.
+
+## Alternatives
+
+In the nearcore the host functions for another pairing-friendly curve BN254 are already implemented[^10]. For some projects[^20] the alternative is just to use the supported curve. However, according to the recent research, this curve contains less than 100 bits of security and is not recommended to use[^13]. Moreover, projects with cross-chain interactions, such as Rainbow Bridge, must use the same curve as in a target protocol, and for Ethereum it is BLS12-381[^3] nowadays. As a result, there is no alternative to using another pairing-friendly curve.
+
+Another alternative is to create one simple host function in nearcore for BLS-signature verification. It was the first suggested solution[^26]. However, this solution is not flexible enough[^28]: (1) projects can use different hash functions; (2) some projects can use G1 subgroup for the public keys, and others – G2; (3) the specification for Ethereum 2.0 continue to be in the draft and details can change, (4) we have one big function instead of a more diverse and flexible set of functions (inspired by EIP-2537's precompiles).
+
+The next alternative is to execute BLS12-381 operations off-chain. In that case, the applications which are using the BLS curve will not be trustless anymore.
+
+## Future possibilities
+
+In the future, it is possible to support work with other curves, not only BLS12-381. In Ethereum, before EIP-2537[^15], EIP-1962 was proposed[^27]. In EIP-1962 was proposed to implement pairing-friendly elliptic curves in a generic format and support not only BLS curves but many others. However, this proposal wasn't accepted due to its large scope and complexity. I don't think it makes sense to implement every possible curve, but it could be a possible extension.
+
+## Consequences
+
+### Positive
+
+- Projects, which are using BN254 will be able to switch on BLS12-381 curve, and it will improve their security.
+- The trustless cross-chain interactions with blockchains using BLS12-381 in protocols (such as Ethereum 2.0) will become possible.
+
+### Neutral
+
+### Negative
+
+- The appearance of dependence on the library which supports BLS12-381 curves operations.
```

Comment:
Also add the fact that we must maintain a bls12-381 implementation forever, even when it becomes broken.

(This is one thing often reproached to OpenPGP, especially its support of 3DES or old mac schemes)

Reply:
There is an important difference between how cryptographic systems were designed in 1990s vs today.
In 1990s, it was normal to pack many different algorithms into a system so that if some of them turn out to be broken, others hopefully remain secure. Protocols like TLS and OpenPGP hail from that era.
Nowadays, the best practice is to only support one or a very small number of algorithms, to make sure that none of them is broken. It turned out that even a single broken algorithm makes the entire system much weaker, because an attacker might be able to make the system fall back to using an insecure algorithm, or the user might be using it accidentally. Modern systems like [libsodium](https://en.wikipedia.org/wiki/Libsodium), [AGE](https://github.com/FiloSottile/age), [Double Ratchet](https://en.wikipedia.org/wiki/Double_Ratchet_Algorithm) and others follow this principle.
NEAR has to implement multiple cryptographic primitives in order to interoperate with other systems (such as Ethereum), but we evaluate each one separately, so hopefully we won't end up in OpenPGP situation.

Reply:
I totally agree that we hopefully won’t end up in a bad situation (regardless of whether the OpenPGP situation is bad), and this is why I’m not saying this negative point should be blocking. But it should be explicitly listed, because we ought to be lucid about what the drawbacks of our NEPs are.

Reply:
@Ekleog-NEAR Sure, I will add this disadvantage

Reply:
fixed: https://github.com/near/NEPs/pull/488/commits/6e9c7b0728791b93fea92fb18ea13f54c6acc496